### PR TITLE
NDData show units

### LINKED ISSUE #11541: Print unit in NDData
### Description

While discussing #11210 with @mhvk, I noticed that printing NDData with a set unit does not show the unit. While not a bug, my expectation is that the unit would be shown. I don't see any Issues or PRs on this... so maybe it's an intentional design decision?

### Additional context
E.g

```python
>>> from astropy.nddata import NDData; print(NDData([1, 2, 3], unit=u.km))
NDData([1, 2, 3])
```

### PR DESCRIPTION
**Pull request by nstarman**
### Description

Now NDData with units shows the units on ``str()`` and ``repr()``. Importantly, ``repr`` maintains its current level of fidelity to ``eval(repr(x)) == x`` (I add a check for this to the tests)

```
>>> x = NDData([1, 2, 3], unit="km")
>>> print(repr(x))
NDData([1, 2, 3], unit='km')

>>> print(str(x))
[1 2 3] km
```

@mhvk @saimn @mwcraig 

Fixes #11541 

Followup ideas:
- ``eval(repr(x)) == x`` doesn't actually work for NDData, with or without units, even if they are identical attribute-by-attribute. Modding ``__eq__`` would be good.
- show other attributes in string representations: mask, uncertainty, ...

### CODE REVIEW

File: astropy/nddata/tests/test_nddata.py
Comment (pllim) [inline astropy/nddata/tests/test_nddata.py:427]: Might also want to test for more complex using like "erg cm^-2 s^-1 A^-1" 😬

Comment (nstarman) [reply astropy/nddata/tests/test_nddata.py:427]: Thanks! will do.

Comment (nstarman) [reply astropy/nddata/tests/test_nddata.py:427]: Done!

File: astropy/nddata/nddata.py
Comment (saimn) [inline astropy/nddata/nddata.py:247]: Using a f-string would be clearer, e.g. `f'{self.__class__.__name__}({data}{unit})`

**Comment by saimn** [comment]
:+1: for me, but @mwcraig should approve as the nddata maintainer.

File: astropy/nddata/nddata.py
Comment (nstarman) [reply astropy/nddata/nddata.py:247]: Agreed!

Comment (nstarman) [reply astropy/nddata/nddata.py]: I missed it, but the prefix is necessary fo indent levels and this makes the .join method the most succinct.

Comment (saimn) [reply astropy/nddata/nddata.py]: Oh indeed, I  missed that too.

**Review by mwcraig — APPROVED** [review]
thanks, @nstarman, this looks great!

**Comment by mwcraig** [comment]
> `eval(repr(x)) == x` doesn't actually work for NDData, with or without units, even if they are identical attribute-by-attribute. Modding `__eq__` would be good.

I opened an issue for this at https://github.com/astropy/astropy/issues/11648

**Pull request merged**
